tools/pykernel/_src/kernel_ast.py

```diff
@@ -892,15 +834,178 @@ def populate_list(arr, _idx=[]):
 
         return var
 
+    def visit_Tuple(self, node):
+        return tuple(map(self.visit, node.elts))
+
     # Literals
     def visit_Constant(self, node):
         as_attr = getattr(node, "_ttkernel_as_attr", False)
         op_constructor = IntegerAttr.get if as_attr else arith.ConstantOp
-        if isinstance(node.value, bool):
+        if callable(as_attr):
+            return as_attr(node)
+        elif isinstance(node.value, bool):
             return op_constructor(IntegerType.get_signless(1, self.ctx), node.value)
         elif isinstance(node.value, int):
             return op_constructor(IntegerType.get_signless(32, self.ctx), node.value)
         else:
             raise NotImplementedError(
                 f"constant type {type(node.value).__name__} not implemented"
             )
+
+
+class TTKernelCompiler(TTCompilerBase):
```

Comment:
I am not looking at this line by line on the premise that it is largely the same -- just relocated; if there are specific parts that changed that should be checked, please point them out.